File: Samples VB.NET/Advanced/CategoryDesigner/TranslationViewModel.vb
Imports System
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports System.Collections.Specialized
Imports System.ComponentModel
Imports System.Data
Imports System.Globalization
Imports System.Linq
Imports System.Text
Imports System.Windows
Imports System.Windows.Controls

Namespace Therefore.Samples
	Public Class TranslationViewModel
		Implements INotifyPropertyChanged
		Private _languages As EventDictionary(Of Integer, CultureInfo)
		Private _translationlist As ObservableCollection(Of TranslationItem)

		Public Property Translations() As ObservableCollection(Of TranslationItem)
			Get
				Return _translationlist
			End Get
			Set(ByVal value As ObservableCollection(Of TranslationItem))
				If _translationlist IsNot value Then
					_translationlist = value
					OnPropertyChanged("Translations")
				End If
			End Set
		End Property

		Public ReadOnly Property Languages() As EventDictionary(Of Integer, CultureInfo)
			Get
				Return _languages
			End Get
		End Property

		Public Sub UpdatedLanguages()
			OnPropertyChanged("Languages")
		End Sub

		Public Sub SetTranslation(ByVal translations As List(Of TranslationItem), ByVal languages As EventDictionary(Of Integer, CultureInfo))
			_translationlist = New ObservableCollection(Of TranslationItem)(translations)
			_languages = languages
			AddHandler _languages.DictionaryItemRemoved, AddressOf OnLanguageRemoved
		End Sub

		Private Sub OnLanguageRemoved(ByVal sender As Object, ByVal e As DictionaryRemoveItemEventArgs(Of Integer))
			'this needs to be done to not recreate a deleted translation when saving the changes
			For Each item As TranslationItem In _translationlist
				item.Translations.Remove(e.Key)
			Next item
		End Sub

		Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged
		Protected Overridable Sub OnPropertyChanged(ByVal propertyName As String)
			RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(propertyName))
		End Sub
	End Class
End Namespace

File: Samples CSharp/Advanced/CategoryDesigner/TranslationViewModel.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Collections.Specialized;
using System.ComponentModel;
using System.Data;
using System.Globalization;
using System.Linq;
using System.Text;
using System.Windows;
using System.Windows.Controls;

namespace Therefore.Samples
{
    public class TranslationViewModel : INotifyPropertyChanged
    {
        private EventDictionary<int, CultureInfo> _languages;
        private ObservableCollection<TranslationItem> _translationlist;

        public ObservableCollection<TranslationItem> Translations
        {
            get { return _translationlist; }
            set
            {
                if(_translationlist != value)
                {
                    _translationlist = value;
                    OnPropertyChanged("Translations");
                }
            }
        }
        
        public EventDictionary<int, CultureInfo> Languages
        {
            get
            {
                return _languages;
            }
        }

        public void UpdatedLanguages()
        {
            OnPropertyChanged("Languages");
        }
        
        public void SetTranslation(List<TranslationItem> translations, EventDictionary<int, CultureInfo> languages)
        {
            _translationlist = new ObservableCollection<TranslationItem>(translations);
            _languages = languages;
            _languages.DictionaryItemRemoved += OnLanguageRemoved;
        }

        private void OnLanguageRemoved(object sender, DictionaryRemoveItemEventArgs<int> e)
        {
            //this needs to be done to not recreate a deleted translation when saving the changes
            foreach (TranslationItem item in _translationlist)
                item.Translations.Remove(e.Key);
        }

        public event PropertyChangedEventHandler PropertyChanged;
        protected virtual void OnPropertyChanged(string propertyName)
        {
            if (this.PropertyChanged != null)
                this.PropertyChanged(this, new PropertyChangedEventArgs(propertyName));
        }
    }
}